storey/flow.py

```diff
@@ -153,6 +156,85 @@ def raise_error_or_return_termination_result():
         return FlowController(self._emit, raise_error_or_return_termination_result)
 
 
+class AsyncAwaitableResult:
+    def __init__(self):
+        self._q = asyncio.Queue(1)
+
+    async def await_result(self):
+        return await self._q.get()
+
+    async def _set_result(self, element):
+        await self._q.put(element)
+
+    async def _set_error(self, ex):
+        await self._set_result(ex)
+
+
+class AsyncFlowController:
+    def __init__(self, emit_fn, loop_task):
+        self._emit_fn = emit_fn
+        self._loop_task = loop_task
+
+    async def emit(self, element, key=None, event_time=None, await_result=False):
+        awaitable = None
+        if event_time is None:
+            event_time = time.time()
+        if await_result:
+            awaitable = AsyncAwaitableResult()
+        await self._emit_fn(Event(element, key, event_time, awaitable))
+        if await_result:
+            result = await awaitable.await_result()
+            if isinstance(result, BaseException):
+                raise result
+            return result
+
+    async def terminate(self):
+        await self._emit_fn(_termination_obj)
+
+    async def await_termination(self):
+        return await self._loop_task
+
+
+class AsyncSource(Flow):
+    def __init__(self, buffer_size=1, **kwargs):
+        super().__init__(**kwargs)
+        if buffer_size <= 0:
+            raise ValueError('Buffer size must be positive')
+        print(f'AsyncSource: asyncio.get_running_loop()={asyncio.get_running_loop()}')
+        self._q = asyncio.Queue(buffer_size, loop=asyncio.get_running_loop())
+        self._ex = None
+
+    async def _run_loop(self):
+        while True:
+            event = await self._q.get()
+            try:
+                termination_result = await self._do_downstream(event)
+                if event is _termination_obj:
+                    return termination_result
+            except BaseException as ex:
+                self._ex = ex
+                if event.awaitable_result:
+                    awaitable = event.awaitable_result._set_error(ex)
+                    if awaitable:
+                        await awaitable
+                if not self._q.empty():
+                    await self._q.get()
```

Comment:
can't you have multiple requests in _q?

Reply:
So that the client can unblock if it's stuck on put().